compiler/tests/fixtures/data_copying_stress.fe

```diff
@@ -0,0 +1,45 @@
+contract Foo:
+    my_string: string42
+    my_other_string: string42
+
+    my_u256: u256
+    my_other_u256: u256
+
+    event MyEvent:
+        my_string: string42
+        my_u256: u256
+
+    pub def set_my_vals(
+        my_string: string42,
+        my_other_string: string42,
+        my_u256: u256,
+        my_other_u256: u256
+    ):
+        self.my_string = my_string
+        self.my_other_string = my_other_string
+        self.my_u256 = my_u256
+        self.my_other_u256 = my_other_u256
+
+    pub def set_to_my_other_vals():
+        self.my_string = self.my_other_string
```

Comment:
I'm confused about what is happening here. Are we *copying* because strings are currently copied by default? Or are we setting `self.my_string` to another storage reference?

My first reaction was: Ah, we are probably copying because strings are set to implicitly copy but then further down I see this code:

```
pub def emit_my_event():
        self.emit_my_event_internal(
            self.my_string.to_mem(),
            self.my_u256.to_mem()
        )
```

...which makes me think they aren't implicitly copied and need `.to_mem()` to explicitly copy them. Can you clarify what happens here?

To be honest, the more I think about it, the more I'm leaning towards **not implicitly copying** for any variable that is in storage no matter its type. This may feel a little awkward for things like numerics where people are just so used to having them copied implicitly but at the same time it  makes the semantics very very clear.

Because strings in Fe are currently not much different from arrays. Both have a fixed max length but can occupy any size between 0 and the max length. So you can have a `string10000` and it would be hard to explain why we implicitly copy it but demand `to_nem` for an `u256[10000]`.

Reply:
> ...which makes me think they aren't implicitly copied and need .to_mem() to explicitly copy them. Can you clarify what happens here?

This is a memory to storage copy: `self.my_string = self.my_other_string.to_mem()`

This is a storage to storage copy: `self.my_string = self.my_other_string`

Reply:
> Are we copying because strings are currently copied by default?

Strings, arrays, and any other reference types will be treated the same in terms of assignments.

Reply:
> To be honest, the more I think about it, the more I'm leaning towards not implicitly copying for any variable that is in storage no matter its type. This may feel a little awkward for things like numerics where people are just so used to having them copied implicitly but at the same time it makes the semantics very very clear.

This is to say that something like

```
my_num: u256 = self.my_nums[42]
```

would not automatically load the value stored in the pointer at index 42 onto the stack? So we would instead need another attribute function to load this onto the stack?


```
my_num: u256 = self.my_nums[42].to_stack()
```

I think we should stick with the conventional behavior of loading onto the stack when indexing a value type.

Reply:
>I think we should stick with the conventional behavior of loading onto the stack when assigning a value type.

Ok, but isn't it confusing then that I have to call `to_mem()` on `self.my_u256` here. 

```
pub def emit_my_event():
        self.emit_my_event_internal(
            self.my_string.to_mem(),
            self.my_u256.to_mem()
        )
```

Why wouldn't that cause an implicit copy directly?

Reply:
Btw, this is my mental model of how I assume things to work now:


|           |                      Mem  |                               Storage |
|---------|---------------------------|-------------------------------------|
| Mem  | `some = val.clone()`  | `some = self.val.to_mem()`  |
| Storage  |  `self.some = val.clone()` |  `self.some = self.val` |

For numeric

|           |          Mem  |            Storage   |
|---------|-----------------|-----------------------|
| Mem  | `some = val`  | `some = self.val`  |
| Storage  |  `self.some = val` |  `self.some = self.val` |

Reply:
I agree with the concern @cburgdorf in terms of semantics but for the user-friendliness, I will vote for implicit conversion to memory as this is something every code used to, Almost every other language does that. We may need to communicate our message more strongly to our developer community for some exceptions like a string in this case.

For the event case, I do agree with you @cburgdorf that it should be implicit conversion as with standard semantics. I haven't checked the code thoroughly to find gotchas in it but I am assuming @g-r-a-n-t is very well aware of this behavior and if we want we can log it in our backlog.